candle-transformers/src/models/z_image/transformer.rs

```diff
@@ -0,0 +1,1087 @@
+        let device = hidden_states.device();
+
+        // Cross-platform attention dispatch
+        let context = self.attention_dispatch(&q, &k, &v, attention_mask, scale, device)?;
+
+        // Reshape back: (B, n_heads, seq_len, head_dim) -> (B, seq_len, dim)
+        let context = context.transpose(1, 2)?.reshape((b, seq_len, ()))?;
+
+        context.apply(&self.to_out)
+    }
+
+    /// Cross-platform attention dispatch
+    fn attention_dispatch(
+        &self,
+        q: &Tensor,
+        k: &Tensor,
+        v: &Tensor,
+        mask: Option<&Tensor>,
+        scale: f64,
+        device: &Device,
+    ) -> Result<Tensor> {
+        // If acceleration disabled, use basic implementation
+        if !self.use_accelerated_attn {
+            return self.attention_basic(q, k, v, mask, scale);
+        }
+
+        // Platform dispatch: prefer optimal implementation per platform
+        if device.is_cuda() {
+            self.attention_cuda(q, k, v, mask, scale)
+        } else if device.is_metal() {
+            self.attention_metal(q, k, v, mask, scale)
+        } else {
+            // CPU fallback
+            self.attention_basic(q, k, v, mask, scale)
+        }
+    }
+
+    /// CUDA: Use Flash Attention
+    #[allow(unused_variables)]
+    fn attention_cuda(
+        &self,
+        q: &Tensor,
+        k: &Tensor,
+        v: &Tensor,
+        mask: Option<&Tensor>,
+        scale: f64,
+    ) -> Result<Tensor> {
+        #[cfg(feature = "flash-attn")]
+        {
+            // flash_attn does not directly support custom mask
+            // Fallback to basic implementation when mask is present
+            if mask.is_some() {
+                return self.attention_basic(q, k, v, mask, scale);
+            }
+
+            // flash_attn input format: (batch, seq_len, num_heads, head_size)
+            // Current format: (batch, num_heads, seq_len, head_size)
+            let q = q.transpose(1, 2)?;
+            let k = k.transpose(1, 2)?;
+            let v = v.transpose(1, 2)?;
+
+            let result = flash_attn(&q, &k, &v, scale as f32, false)?;
+            result.transpose(1, 2)
+        }
+
+        #[cfg(not(feature = "flash-attn"))]
+        {
+            // flash-attn not compiled, fallback to basic
+            self.attention_basic(q, k, v, mask, scale)
+        }
+    }
+
+    /// Metal: Use fused SDPA kernel
+    fn attention_metal(
+        &self,
+        q: &Tensor,
+        k: &Tensor,
+        v: &Tensor,
+        mask: Option<&Tensor>,
+        scale: f64,
+    ) -> Result<Tensor> {
+        // Prepare SDPA format mask
+        let sdpa_mask = self.prepare_sdpa_mask(mask, q)?;
+
+        // candle_nn::ops::sdpa
+        // Input format: (bs, qhead, seq, hidden) - matches current format
+        // Supports: BF16/F16/F32, head_dim=128
+        candle_nn::ops::sdpa(q, k, v, sdpa_mask.as_ref(), false, scale as f32, 1.0)
+    }
+
+    /// CPU: Basic step-by-step implementation
```

Comment:
This is also used by cuda when compiled without flash attention
```suggestion
    /// Fallback implementation
```